doc/monads/monads.lean

```diff
@@ -0,0 +1,307 @@
+is `Nat → Option Float` The **bind operator resolves** these to an `Option Float`. Then you pass this
+similarly through the bind operator to `optionFunc3`, resulting in the final type, `Option (List Nat)`.
+
+Your functions will not always combine so cleanly though. This is where `do` notation comes into play.
+This notation allows you to write monadic operations one after another, line-by-line. It almost makes
+your code look like imperative programming. You can rewrite the above as:
+-/
+
+def runOptionFuncsDo (input : String) : Option (List Nat) := do
+  let i ← optionFunc1 input
+  let f ← optionFunc2 i
+  optionFunc3 f
+
+#eval runOptionFuncsDo "big" -- some [9, 10]
+/-!
+
+Note you can use `<-` or the nice unicode symbol `←` which you can type into VS code by typing these
+characters `\l `.  When you type the final space, `\l` is replaced with `←`.
+
+The `←` operator is special. It effectively unwraps the value on the right-hand side from the monad.
+This means the value `i` has type `Nat`, _even though_ the result of `optionFunc1` is `Option Nat`.
+The bind operation happens under the hood. If the function returns `none`, then the entire
+`runOptionFuncsDo` function will return `none`. Observe that you do not unwrap the final line of the
+computation. The function result is `Option (List Nat)` which matches what `optionFunc3` returns.
+
+At first glance, this looks more complicated than the bind example. However, it gives you a lot more
+flexibility, like mixing monadic and non-monadic statements, using if then/else structures with
+their own local do blocks and so on. It is particularly helpful when one monadic function depends on
+multiple previous functions.
+
+## Example using List
+
+You can easily make `List` into a monad with the following, since List already provides an
+implementation of `pure` and `bind`.
+
+-/
+instance : Monad List  where
+  pure := List.pure
+  bind := List.bind
+/-!
+
+Like you saw with the applicative `seq` operator, the `bind` operator applies the given function
+to every element of the list.  It is useful to look at the bind implementation for List:
+
+-/
+open List
+def bind (a : List α) (b : α → List β) : List β := join (map b a)
+/-!
+
+So `Functor.map` is used to apply the function `b` to every element of `a` but this would
+return a whole bunch of little lists, so `join` is used to turn those back into a single list.
+
+Here's an example where you use `bind` to convert a list of strings into a combined list of chars:
+
+-/
+
+#eval "apple".toList  -- ['a', 'p', 'p', 'l', 'e']
+
+#eval ["apple", "orange"] >>= String.toList
+-- ['a', 'p', 'p', 'l', 'e', 'o', 'r', 'a', 'n', 'g', 'e']
+/-!
+List also provides some functions that are designed to operate in the context of a monad.
+These methods end in upper case M like `anyM` below:
+-/
+
+def hasSomeItemGreaterThan (x : List Nat) (n : Nat): Option Bool := do
```

Comment:
This example is strange. `Option` is not necessary here, and one can write more compact pure code.
```lean
def hasSomeItemGreaterThan' (x : List Nat) (n : Nat) : Bool :=
  x.any λ a => if a > n then true else false 
```
BTW, note that a novice reader may also get confused by how this example even type checks in Lean. Lean is injecting a coercion (`some` in this case) at `x`.
Could you please find a better example?

Reply:
Can you suggest a better example of `anyM` or any other monadic function provided by List ?

Reply:
If I remember correctly, you have defined only two monads in this section `Option` and `List`. If you want to keep the example with `anyM`, I think it is better to move it to the `Reader` section. Example
```
structure Env where 
  min : Nat
  max : Nat

def hasSomeItemOutsideRange (x : List Nat) : ReaderM Env Bool :=
  x.anyM fun a => do let env ← read; pure (a < env.min || a > env.max)
```

Reply:
Thanks, I like the idea of requiring a more interesting monad context like ReaderM, or Except perhaps?  Problem with hasSomeItemOutsideRange is I would normally just lift `let env ← read` outside and call x.any.

How about this one:

```lean
def validateList (x : List Nat) (n : Nat): Except String Bool := do
  x.anyM (λ a => if a > n then do throw "illegal value found in list" else pure true)

#eval validateList [1, 2, 5, 3, 8] 10 -- some true
```

Reply:
I'll move this to except.lean.

Reply:
> Thanks, I like the idea of requiring a more interesting monad context like ReaderM, or Except perhaps? Problem with hasSomeItemOutsideRange is I would normally just lift let env ← read outside and call x.any.

Yes, you could do that. We can create a realistic example by using 
```lean
def inRange (x : Nat) :  ReaderM Env Bool := do
  let env ← read
  pure (a < env.min || a > env.max)

def hasSomeItemOutsideRange (x : List Nat) : ReaderM Env Bool :=
  x.anyM inRange
```
You will find this kind of pattern in many places in our code base.

Reply:
The following example is not very realistic since the nested function fails or returns true. That is, it never returns false. It feels weird to use `anyM`.
```lean
def validateList (x : List Nat) (n : Nat): Except String Bool := do
  x.anyM (λ a => if a > n then do throw "illegal value found in list" else pure true)
```
A more realistic option is to use
```lean
def validateList (x : List Nat) (n : Nat): Except String Unit :=
  x.forM fun a => do 
    if a > n then throw "illegal value found in list"
```

Reply:
Ok, that's cool, I'll use this one, thanks.

Reply:
Fixed in https://github.com/leanprover/lean4/pull/1505/commits/40da0e10b21be41da5e54fdf709a14ff447abde6